velox/functions/prestosql/types/JsonType.cpp

```diff
@@ -856,30 +885,55 @@ struct CastFromJsonTypedImpl {
         }
       } else {
         SIMDJSON_ASSIGN_OR_RAISE(auto object, value.get_object());
-        folly::F14FastMap<std::string, simdjson::ondemand::value> lowerCaseKeys(
-            object.count_fields());
+
+        // TODO Populate this mapping once, not per-row.
+        // Mapping from lower-case field names of the target RowType to their
+        // indices.
+        bool allFieldsAreAscii = true;
+        for (auto i = 0; i < rowType.size(); ++i) {
+          allFieldsAreAscii &= isAscii(rowType.nameOf(i));
+        }
+
+        auto fieldIndices = makeFieldIndicesMap(rowType, allFieldsAreAscii);
+
+        // Keep track of which fields have been found in the JSON object.
+        std::vector<bool> foundIndices(rowType.size(), false);
+
         std::string key;
         for (auto fieldResult : object) {
           SIMDJSON_ASSIGN_OR_RAISE(auto field, fieldResult);
           if (!field.value().is_null()) {
             SIMDJSON_ASSIGN_OR_RAISE(key, field.unescaped_key(true));
-            boost::algorithm::to_lower(key);
-            lowerCaseKeys[key] = field.value();
+            std::string lowerCaseKey = key;
+
+            // boost::algorithm::to_lower is very slow. Use much faster
+            // folly::toLowerAscii if possible.
+            if (allFieldsAreAscii) {
+              folly::toLowerAscii(lowerCaseKey);
```

Comment:
Is `toLowerAscii` safe to be called on non-ascii input?  If not we may need to check if `lowerCaseKey` is ascii as well

Reply:
I think it is: "Leaves all other characters unchanged, including those with the 0x80 bit set."

```
/**
 * Convert ascii to lowercase, in-place.
 *
 * Leaves all other characters unchanged, including those with the 0x80
 * bit set.
 * @param str String to convert
 * @param length Length of str, in bytes
 */
void toLowerAscii(char* str, size_t length);

inline void toLowerAscii(std::string& str) {
  // str[0] is legal also if the string is empty.
  toLowerAscii(&str[0], str.size());
}
```